### ISSUE TITLE
Procedure 'Node Upgrade Using the SUSE Distribution Migration System' does not work.

### ISSUE THREAD

**Issue by mvangool** [SUSE/doc-ses#715]
**Link To Section**
https://documentation.suse.com/ses/6/html/ses-all/cha-ceph-upgrade.html#upgrade-one-node-auto

**Summary**
The procedure (6.8.2 Node Upgrade Using the SUSE Distribution Migration System ) as it is currently doesn't work for systems registered against SCC.
After step 3 the migration will start and fails with the following message:

[INFO ]: 09:11:13 | Running prepare service
[INFO ]: 09:11:13 | Bind mounting /etc/zypp
[INFO ]: 09:11:13 | Bind mounting /usr/lib/zypp/plugins
[INFO ]: 09:11:13 | Mounting kernel file systems inside /system-root
[INFO ]: 09:11:14 | Creating backup of Product data
[INFO ]: 09:11:14 | Updating Base Product to be suitable for migration
[INFO ]: 09:11:14 | Running migrate service

Executing 'zypper --root /system-root --non-interactive --gpg-auto-import-keys refresh'

Repository 'SUSE-Enterprise-Storage-5-Pool' is up to date.
Building repository 'SUSE-Enterprise-Storage-5-Pool' cache [....done]
Repository 'SUSE-Enterprise-Storage-5-Updates' is up to date.
Building repository 'SUSE-Enterprise-Storage-5-Updates' cache [....done]
Repository 'SLES12-SP3-Pool' is up to date.
Building repository 'SLES12-SP3-Pool' cache [....done]
Repository 'SLES12-SP3-Updates' is up to date.
Building repository 'SLES12-SP3-Updates' cache [....done]
All repositories have been refreshed.
Can't get available migrations from server: SUSE::Connect::ApiError: The requested products 'SUSE Linux Enterprise Server LTSS 12 SP3 x86_64' have not been activated on the system.
'/usr/lib/zypper/commands/zypper-migration' exited with status 1
[ERROR ]: 09:11:36 | migrate service failed with bash: stderr: (no output on stderr), stdout: (no output on stdout)
[INFO ]: 09:11:36 | Running grub setup service
[INFO ]: 09:11:44 | Uninstalling migration:
>>>>

**Steps To Reproduce**
This is for the scenario where a system is registered with SCC directly, so following steps 1, 2a, 3 will reproduce the error.

**Notes/Possible Solution**
The procedure is failing because the following 3 packages are installed:
SLES-LTSS
sles-ltss-release
sles-ltss-release-POOL

Adding an additional step (removing these 3 packages) to the procedure will 'fix' the problem.

**Comment by asettle** [comment]
@swiftgist - we can easily fix this in the documentation as a workaround, but could you please review this from a technical POV? Just want to ensure there isn't something we could be fixing here to make this suitable for SCC users.

**Comment by swiftgist** [comment]
@tserong is more knowledgeable about SCC.  I do not believe I can actually create the environment to cause the original issue.

**Comment by asettle** [comment]
@tserong any help on this one? :) thanks!

**Comment by tserong** [comment]
I think the key issue here is highlighted in the error message "The requested products 'SUSE Linux Enterprise Server LTSS 12 SP3 x86_64' have not been activated on the system". I _assume_ that what's happening here is SCC is spitting the dummy because it thinks (rightly or wrongly) that the LTSS product isn't activated on that system. There should thus be two ways to fix the problem; one is for the user to (somehow) activate the LTSS product in SCC, the other is to remove those LTSS product packages from the system. But, I'm not sure this should necessarily be addressed in the SES upgrade documentation, as AIUI this type of problem could occur for _any_ SLE user running `zypper migration`, even outside the context of the DMS, if systems are somehow improperly registered with SCC.

**Comment by asettle** [comment]
@mvangool any chance you could test this solution?

**Comment by asettle** [comment]
@tserong so I've spoken to Mark, and what it boils down to is that the instructions still don't work for this use case. Mark's agreed to write up a TID, so it will be easily searchable for our customers. We could add a note that says something like, "If you are doing this, then you need to manually remove the packages", but I don't want to cause confusion. Do you have any thoughts?

**Comment by tserong** [comment]
Adding a note probably makes sense, but I'd try to make it as generic as possible...  In that section, it already says "If the node being upgraded is registered with a repository staging system such as SCC, SMT, RMT, or SUSE Manager [...set use_zypper_migration: true...]. Perhaps add something like "in this case, zypper will try to migrate all products installed on the system, if any haven't been activated the migration will fail - to fix this, either ensure those products are activated on the system, or alternately, remove the offending product packages"? Or does that just suck?

**Comment by asettle** [comment]
On it, thanks @tserong - @mvangool let me know if you have any concerns about that solution :)

**Comment by timscotland** [comment]
I have just finished resolving a case with a test system (SES5) that we are upgrading.  The source for the original install of the servers was SMT12 that has a local LTSS repository (that might be a blind ally).

The test system was installed using a variation of the Martin & Frieder install system (https://github.com/Martin-Weiss/cif) that I have used for 3 years now and all worked well.   When doing the DMS upgrade I too recieved an error 

_All repositories have been refreshed.
Can't get available migrations from server: SUSE::Connect::ApiError: The requested products 'SUSE Linux Enterprise Server LTSS 12 SP3 x86_64' are not activated on the system.
'/usr/lib/zypper/commands/zypper-migration' exited with status 1
[ERROR ]: 16:26:16 | migrate service failed with bash: stderr: (no output on stderr), stdout: (no output on stdout)
[INFO ]: 16:26:16 | Running grub setup service
[INFO ]: 16:26:19 | Uninstalling migration:_

As can be guessed there were no LTSS repos enabled on the system
t-ses-1:~ # zypper lr
Repository priorities are without effect. All enabled repositories share the same priority.

# | Alias                                                        | Name                              | Enabled | GPG Check | Refresh
--+--------------------------------------------------------------+-----------------------------------+---------+-----------+--------
1 | SLES12-SP3-12.3-0                                            | SLES12-SP3-12.3-0                 | Yes     | (r ) Yes  | Yes
2 | SMT-http_smt12_lshtm_ac_uk:SLES12-SP3-Pool                   | SLES12-SP3-Pool                   | Yes     | (r ) Yes  | No
3 | SMT-http_smt12_lshtm_ac_uk:SLES12-SP3-Updates                | SLES12-SP3-Updates                | Yes     | (r ) Yes  | Yes
4 | SMT-http_smt12_lshtm_ac_uk:SUSE-Enterprise-Storage-5-Pool    | SUSE-Enterprise-Storage-5-Pool    | Yes     | (r ) Yes  | No
5 | SMT-http_smt12_lshtm_ac_uk:SUSE-Enterprise-Storage-5-Updates | SUSE-Enterprise-Storage-5-Updates | Yes     | (r ) Yes  | Yes
t-ses-1:~ #


My fix was based on a TID that referenced a similar issue with a SAP system 
https://www.suse.com/support/kb/doc/?id=000019394

In my case there was just a spurious SLES-LTSS.prod entry in /etc/products.d   I removed that file and re ran the DMS and all worked (on all the 6 test nodes I am working with)

Rolling back to an earlier snapshot I can see that the three entries that Mark mentioned are installed, but the removal of /etc/products.d/SLES-LTSS.prod was all that was required.   

Is the correct documentation alteration to check .etc/products.d and see what the correlation is between the entries there and zypper lr?

Less prone to error and easier to write.

**Comment by timscotland** [comment]
Further experimentation. Build a new SLES12.3 server fully patched - no LTSS items added
Add SES5 (as an add-on product) and the two LTSS items are added as part of the SES5 installation and the SLES-LTSS.prod is added to /etc/products.d

Not tested with a slipstreamed install that included SES from the outset.

So 100% a bug but with SES5 going EOL in January there will be a lot of customers that have the issue.  

I would put forward that part of the upgrade process is to check /etc/products.d  for an SLES-LTSS.prod file and delete said file if found.

**Comment by tserong** [comment]
> Add SES5 (as an add-on product) and the two LTSS items are added as part of the SES5 installation and the SLES-LTSS.prod is added to /etc/products.d

OK, I didn't expect *that* :-/

> I would put forward that part of the upgrade process is to check /etc/products.d for an SLES-LTSS.prod file and delete said file if found.

Given your findings, that's probably the simplest thing to do/explain.

**Comment by asettle** [comment]
@timscotland @mvangool - so, it's taken us a little time to figure this out. Especially because: why is the LTSS product being installed without either of you asking for it to be installed.

After some digging and checking, according to https://www.suse.com/lifecycle/, general support for SLE 12 SP3 ended on 30 Jun 2019, so this SLE version is only available under LTSS conditions (till 30 Jun 2022). 

This means you _have_ to activate the subscription, rather than remove it. This is something we can add in the docs as a step/call-out note :)

Thanks so much for bringing this to our attention. It took a little longer than expected.

**Comment by timscotland** [comment]
@asettle  - Just to be clear, the LTSS element is not part of the repositories listed, and it's 100% not supported as part of a SES5 system

What does seem to be happening is that a patch on the SES5 channel is adding this (almost certainly because of the SLES12 SP3 being EOL) and that is causing the problem (Installing 2 patched and adding the file entry in /etc/products.d)

The question I have that remains is that if you install SLES12.3 & SES5 concurrently does it have the same result of installing first SLES12.3 and then subsequently adding SES6?  Or is it just a case of saying "in some cases" and leaving it at that?

Does it cause harm other than breaking the migration tool?  - not that I can see. So either just removing the 2 files or the SLES-LTSS.prod works.

**Comment by asettle** [comment]
> @asettle - Just to be clear, the LTSS element is not part of the repositories listed, and it's 100% not supported as part of a SES5 system
> 
> What does seem to be happening is that a patch on the SES5 channel is adding this (almost certainly because of the SLES12 SP3 being EOL) and that is causing the problem (Installing 2 patched and adding the file entry in /etc/products.d)
> 
> The question I have that remains is that if you install SLES12.3 & SES5 concurrently does it have the same result of installing first SLES12.3 and then subsequently adding SES5? Or is it just a case of saying "in some cases" and leaving it at that?
> 
> Does it cause harm other than breaking the migration tool? - not that I can see. So either just removing the 2 files or the SLES-LTSS.prod works.
> 
> The first part of this TID explains what might be happening https://www.suse.com/support/kb/doc/?id=000019394

Looking into it.

**Comment by asettle** [comment]
Providing context for future: SES5 is based on SLE-12-SP3, but SLE-12-SP3 went out of maintenance before SES5 did. This gave rise to the situation where customers have a valid product, SES5, but requires an invalid OS, SLE-12-SP3, to run. Just gifting these customers an LTSS subscription was one option, but it was decided to send all LTSS updates to the SES5 channel.

Alright, after a little internal chat, and the fact that the EOL of SES 5 is January, we're going to add the following to the upgrade 5 -> 6 docs:

`Before starting the upgrade, make sure none of the nodes have SLE-12-SP3 LTSS enabled. `

**Comment by timscotland** [comment]
Makes total sense but this is where I hit the problem.

zypper -lr shows no repo, but the migration failed.   This is what 99.9% of folk to ensure that no LTSS is enabled, but the patched are still installed and that the migration fails.

A secondary check of /etc/products.d is required to ensure the migration works.  May I suggest 

Before starting the upgrade, make sure none of the nodes have SLE-12-SP3 LTSS enabled and that SLES-LTSS.prod is not removed if found /etc/products.d

**Comment by smithfarm** [comment]
@timscotland I'm not sure `zypper lr` is the right way to find out if LTSS is enabled. Shouldn't you do something like 

    SUSEConnect --status-text

to find out that information?

**Comment by smithfarm** [comment]
> The question I have that remains is that if you install SLES12.3 & SES5 concurrently does it have
>  the same result of installing first SLES12.3 and then subsequently adding SES5? 

I don't think it's possible to install SLE-12-SP3 and SES5 concurrently. You always have to install the OS first, and then the SES product on top of that. 

You might be able to *register* SLE-12-SP3 and SES5 concurrently, though.

**Comment by timscotland** [comment]
From the box I built to test this I get the following results.  The first thing I would check is the zypper -lr as you would expect that when you run SUSEConnect it will enable and refresh all of the registered repos.  However you are correct in that while it is not registered it is there...

t-ses-6:~ # SUSEConnect --status-text
Installed Products:
------------------------------------------
  SUSE Linux Enterprise Server 12 SP3
  (SLES/12.3/x86_64)
  Registered
------------------------------------------
  SUSE Enterprise Storage 5
  (ses/5/x86_64)
  Registered
------------------------------------------
  SUSE Linux Enterprise Server 12 SP3 LTSS
  (SLES-LTSS/12.3/x86_64)
  Not Registered
------------------------------------------
t-ses-6:~ # zypper lr
Repository priorities are without effect. All enabled repositories share the same priority.

# | Alias                                                        | Name                              | Enabled | GPG Check | Refresh
--+--------------------------------------------------------------+-----------------------------------+---------+-----------+--------
1 | SLES12-SP3-12.3-0                                            | SLES12-SP3-12.3-0                 | Yes     | (r ) Yes  | Yes
2 | SMT-http_smt12_lshtm_ac_uk:SLES12-SP3-Installer-Updates      | SLES12-SP3-Installer-Updates      | Yes     | (r ) Yes  | Yes
3 | SMT-http_smt12_lshtm_ac_uk:SLES12-SP3-Pool                   | SLES12-SP3-Pool                   | Yes     | (r ) Yes  | No
4 | SMT-http_smt12_lshtm_ac_uk:SLES12-SP3-Updates                | SLES12-SP3-Updates                | Yes     | (r ) Yes  | Yes
5 | SMT-http_smt12_lshtm_ac_uk:SUSE-Enterprise-Storage-5-Pool    | SUSE-Enterprise-Storage-5-Pool    | Yes     | (r ) Yes  | No
6 | SMT-http_smt12_lshtm_ac_uk:SUSE-Enterprise-Storage-5-Updates | SUSE-Enterprise-Storage-5-Updates | Yes     | (r ) Yes  | Yes
t-ses-6:/etc/products.d # ll
-rw-r--r-- 1 root root 2651 Jul 11  2019 SLES-LTSS.prod
-rw-r--r-- 1 root root 2917 May  4  2018 SLES.prod
lrwxrwxrwx 1 root root    9 Dec 11 10:26 baseproduct -> SLES.prod
-rw-r--r-- 1 root root 1487 Jul 23 18:55 ses.prod

**Comment by timscotland** [comment]
> > The question I have that remains is that if you install SLES12.3 & SES5 concurrently does it have
> > the same result of installing first SLES12.3 and then subsequently adding SES5?
> 
> I don't think it's possible to install SLE-12-SP3 and SES5 concurrently. You always have to install the OS first, and then the SES product on top of that.
> 
> You might be able to _register_ SLE-12-SP3 and SES5 concurrently, though.

When you register with an SMT it asks if you want to add any of the available extentions and modules - #2 on the list is SES5
![image](https://user-images.githubusercontent.com/29625600/102352775-17b53180-3fa0-11eb-99c8-f6683ba61a00.png)

Just a quick test this AM shows the same issue there in all it's glory
![image](https://user-images.githubusercontent.com/29625600/102352930-534ffb80-3fa0-11eb-9cd2-28133cee778d.png)

**Comment by smithfarm** [comment]
Bingo! What does `SUSEConnect --status-text` say after you remove `/etc/product.s/SLES-LTSS.prod` ?

**Comment by timscotland** [comment]
t-ses-6:/etc/products.d # SUSEConnect --status-text
Installed Products:
------------------------------------------

  SUSE Enterprise Storage 5
  (ses/5/x86_64)

  Registered

------------------------------------------

**Comment by smithfarm** [comment]
I'm afraid I can't answer that. Presumably, someone from the SCC team could shed more light?

**Comment by holgis** [comment]
I did some testing and found out that the file `/etc/products.d/SLES-LTSS.prod` gets removed when uninstalling the `sles-ltss-release*` packages:

```
e112:~ # zypper rm sles-ltss-release sles-ltss-release-POOL
Loading repository data...
Warning: No repositories defined. Operating only with the installed resolvables. Nothing can be installed.
Reading installed packages...
Resolving package dependencies...

The following 2 packages are going to be REMOVED:
  sles-ltss-release sles-ltss-release-POOL

The following product is going to be REMOVED:
  "SUSE Linux Enterprise Server 12 SP3 LTSS"

2 packages to remove.
After the operation, 2,7 KiB will be freed.
Continue? [y/n/...? shows all options] (y): y
(1/2) Removing sles-ltss-release-12.3-10.5.1.x86_64 ......................[done]
(2/2) Removing sles-ltss-release-POOL-12.3-10.5.1.x86_64 .................[done]
```
So it's enough to just uninstall them - no need to delete the file manually.

Here is my proposal for the docu:
Before starting the upgrade process make sure the packages `sles-ltss-release` and `sles-ltss-release-POOL` are **not** installed. If the are, make sure to uninstall them with `zypper rm -y sles-ltss-release sles-ltss-release-POOL`

On my system there was no package `SLES-LTSS` - someone with that package on the system needs to tell us if that needs to be listen in docu as well!

**Comment by asettle** [comment]
Closing this issue with the addition of the LTSS package instruction. I'll be working on #843 to further improve the SES 6 upgrade docs :) any questions or comments, please make them on that issue. This will now be closed. Thank you all!
